keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,498 @@
+        end
+        kmsplugin->>kubeapiserver: return decrypt response <br/> {"plaintext": "<decrypted DEK>", key_id: "<remote KEK ID>", <br/> "annotations": {"kms.kubernetes.io/local-kek": "<encrypted local KEK>"}}
+    else encrypted local KEK is not in annotations
+        kubeapiserver->>kmsplugin: decrypt request <br/> {"ciphertext": "<encrypted DEK>", observed_key_id: "<key_id gotten as part of EncryptResponse>", <br/> "annotations": {}}
+        kmsplugin->>externalkms: decrypt DEK with remote KEK (same behavior as today)
+        externalkms->>kmsplugin: decrypted DEK
+        kmsplugin->>kubeapiserver: return decrypt response <br/> {"plaintext": "<decrypted DEK>", key_id: "<remote KEK ID>", <br/> "annotations": {}}
+    end
+```
+
+### Test Plan
+
+[ ] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+This section is incomplete and will be updated before the beta milestone.
+
+##### Unit tests
+
+This section is incomplete and will be updated before the beta milestone.
+
+##### Integration tests
+
+This section is incomplete and will be updated before the beta milestone.
+
+##### e2e tests
+
+This section is incomplete and will be updated before the beta milestone.
+
+
+### Graduation Criteria
+
+Since the storage version API is still alpha, this KEP will simply aim to make it possible to have automated rotation when that API is enabled and has been updated to include the new fields.  The rotation feature will first be scoped to a single API server and will not be part of the graduation criteria for this KEP because the storage version API will take time to mature. However, testing of rotation will be part of the graduation criteria to confirm that the right information is being made available to allow for automated rotation when the storage version API graduates.
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit and integration tests completed and enabled
+
+#### Beta
+
+TBD
+
+#### GA
+
+TBD
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+-->
+
+- Feature gate
+  - Feature gate name: `KMSv2`
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: false,
+	LockToDefault: false,
+	PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+No.  The v2 API is new in the v1.25 release.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, via the `KMSv2` feature gate. Disabling this gate without first doing a storage migration to use a different encryption at rest mechanism will result in data loss.
```

Comment:
It's more than data loss right, the server would not understand the encryption configuration and I would expect it to fail to start.

Reply:
We are adding a new `APIVersion` field in the KMS configuration which can be used to indicate `v2` API. When the KMSv2 feature gate is disabled, any kms provider that is configured for v2 in the KMS configuration will not take effect because all the code is behind the feature gate. So, this would only result in data loss because existing encrypted data can't be decrypted but the KMS configuration would still load fine. 

cc @enj @ritazh to add anything I missed here